File: day2/puzzle1.go
package main

import (
	"fmt"
	"os"
)

func main() {
	var intCodes []int
	pointer := 0
	calcOnePointer := 0
	calcTwoPointer := 0
	resultPointer := 0

	file, err := os.Open("./input.txt")
	if err != nil {
		fmt.Println(err.Error())
	}
	defer file.Close()

	for {
	    if n, _ := fmt.Fscanf(file, "%d,", &pointer); n != 1 {break}
	    intCodes = append(intCodes,pointer)
	}

	pointer = 0

	for intCodes[pointer] != 99 {
		if intCodes[pointer] == 1 {
			resultPointer = intCodes[pointer+3]
			if resultPointer <= len(intCodes) {
				calcOnePointer = intCodes[pointer+1]
				calcTwoPointer = intCodes[pointer+2]
				intCodes[resultPointer] = intCodes[calcOnePointer] + intCodes[calcTwoPointer]
			}
			pointer += 4
		} else if intCodes[pointer] == 2 {
			resultPointer = intCodes[pointer+3]
			if resultPointer <= len(intCodes) {
				calcOnePointer = intCodes[pointer+1]
				calcTwoPointer = intCodes[pointer+2]
				intCodes[resultPointer] = intCodes[calcOnePointer] * intCodes[calcTwoPointer]
			}
			pointer += 4
		} else {
			pointer++
		}
	}
	fmt.Println(intCodes)
}

File: day2/puzzle1.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    if let Ok(lines) = read_lines("./rust_input.txt") {
        for line in lines {
            if let Ok(intcode_line) = line {
                let intcodes: Vec<i32> = intcode_line
                    .split(',')
                    .map(|s| s.parse().expect("parse error"))
                    .collect();
                run_intcodes(intcodes);
            }
        }
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn run_intcodes(mut ic: std::vec::Vec<i32>){
    let mut pointer = 0;
    loop{
        if ic[pointer] == 99 {
            break;
        } else if ic[pointer] == 1 {
            let arg1 = ic[pointer+1] as usize;
            let arg2 = ic[pointer+2] as usize;
            let answer = ic[pointer+3] as usize;
            ic[answer] = ic[arg1] + ic[arg2];
            pointer += 4;
        } else if ic[pointer] == 2 {
            let arg1 = ic[pointer+1] as usize;
            let arg2 = ic[pointer+2] as usize;
            let answer = ic[pointer+3] as usize;
            ic[answer] = ic[arg1] * ic[arg2];
            pointer += 4;
        } else {
            println!("Invalid opcode -> {}", ic[pointer]);
            break;
        }
    }
    println!("Final array -> {:?}", ic);
}